early-dev/app.R

```diff
@@ -1,339 +1,219 @@
-        DTHFL = "Death Flag",
-        DCSREAS = "Reason for Study Discontinuation",
-        EOSSTT = "End of Study Status"
-      ) %>%
-      droplevels()
-    teal.data::col_labels(ADSL)[c(names(adsl_labels))] <- adsl_labels
-
-    ADAE <- synthetic_cdisc_data("latest")$adae
-
-    # derive common flags for AEs
-    adae_labels <- teal.data::col_labels(ADAE, fill = FALSE)
-    ADAE <- ADAE %>%
-      mutate_at(c("AESOC", "AEBODSYS", "AEHLT", "AEDECOD", "AETERM", "AELLT"), as.character) %>%
-      mutate(
-        RELFL = ifelse(AEREL == "Y", "Y", "N"),
-        CTC35FL = ifelse(AETOXGR %in% c("3", "4", "5"), "Y", "N"),
-        SERFL = ifelse(AESER == "Y", "Y", "N"),
-        RELSERFL = ifelse(AEREL == "Y" & AESER == "Y", "Y", "N"),
-        AEREL1 = (AEREL == "Y" & ACTARM == "A: Drug X"),
-        AEREL2 = (AEREL == "Y" & ACTARM == "B: Placebo"),
-        ASTDT = as.Date(ASTDTM),
-        AENDT = as.Date(AENDTM)
-      ) %>%
-      teal.data::col_relabel(
-        RELFL = "Related AE",
-        CTC35FL = "Grade >=3 AE",
-        SERFL = "Serious AE",
-        RELSERFL = "Related Serious AE",
-        AEREL1 = "AE related to A: Drug X",
-        AEREL2 = "AE related to B: Placebo",
-        ASTDT = "Analysis Start Date",
-        AENDT = "Analysis End Date",
-        AESOC = "Primary System Organ Class",
-        AEBODSYS = "Body System or Organ Class",
-        AEHLT = "High Level Term",
-        AEDECOD = "Dictionary-Derived Term",
-        AETERM = "Reported Term for the Adverse Event",
-        AELLT = "Lowest Level Term"
-      )
+  cs_paramcd_rsp <- choices_selected(
+    choices = value_choices(data[["ADRS"]], "PARAMCD", "PARAM", subset = c("BESRSPI", "INVET")),
+    selected = "BESRSPI"
+  )
 
-    ADCM <- synthetic_cdisc_data("latest")$adcm
-
-    # process ADCM
-    ADCM <- ADCM %>%
-      mutate(
-        ASTDT = as.Date(ASTDTM),
-        AENDT = as.Date(AENDTM)
-      ) %>%
-      teal.data::col_relabel(
-        ASTDT = "Analysis Start Date",
-        AENDT = "Analysis End Date"
-      )
+  cs_paramcd_tr <- choices_selected(
+    choices = value_choices(data[["ADTR"]], "PARAMCD", "PARAM", subset = "SLDINV"),
+    selected = "SLDINV"
+  )
+}
 
-    ADEX <- synthetic_cdisc_data("latest")$adex
-
-    # process ADEX
-    ADEX <- ADEX %>%
-      mutate(
-        ASTDT = as.Date(ASTDTM),
-        AENDT = as.Date(AENDTM)
-      ) %>%
-      teal.data::col_relabel(
-        ASTDT = "Analysis Start Date",
-        AENDT = "Analysis End Date"
-      )
+## Join Keys
+{
+  get_join_keys(data)["ADTR"] <- c("STUDYID", "USUBJID", "PARAMCD", "AVISIT")
+  get_join_keys(data)["ADTRWF"] <- c("STUDYID", "USUBJID", "PARAMCD", "AVISIT")
+  get_join_keys(data)["ADRSSWIM"] <- c("STUDYID", "USUBJID", "PARAMCD", "AVISIT")
+  get_join_keys(data)["ADTR", "ADSL"] <- c("STUDYID", "USUBJID")
+  get_join_keys(data)["ADTRWF", "ADSL"] <- c("STUDYID", "USUBJID")
+  get_join_keys(data)["ADRSSWIM", "ADSL"] <- c("STUDYID", "USUBJID")
+}
```

Comment:
because empty `cdisc_data()` has been created in the beginning it means that `get_join_keys(data)` before this code block is empty as well. This code glock just adds "missing defaults". I think it might be wise to initialize cdisc_data with default keys:

```
cdisc_data(join_keys = cdisc_join_keys("ADSL", "ADAE", "ADCM", ...))
```
and then add above code block. If we somehow pass full list of join_keys to the `cdisc_data()` function, then `datanames(data)` won't be needed - as this will be concluded from the keys (see default `cdisc_data(datanames = c(...))`)

Reply:
Also, what if the data is not relational?

Reply:
You can do this before and after. We have no asserts which would block this. I think we need to figure out optimal solution by playing with this app code - this is a good playground.

Reply:
What I mean is if your data is not relational, `join_keys` will be empty and you won't be able to determine datanames. This is less of an issue with `cdisc_data` but more so with `teal_data`.

Reply:
> Also, what if the data is not relational?
if it is not relational - then one uses `teal_data()` and avoid to specify `join_keys()` (which stays empty all the time then)

Reply:
But we do need non-empty `@datanames` in `init`.